pkg/kubectl/apis/manifest/v1alpha1/types.go

```diff
@@ -0,0 +1,175 @@
+// A manifest could be either a Base or an Overlay.
+type Manifest struct {
+	metav1.TypeMeta `json:",inline" yaml:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	// TODO: figure out if we need field ManifestVersion.
+	// See: https://github.com/kubernetes/kubernetes/pull/52570/files/3eea91793dfbc3fdb0799589fac3790c4cde58a4#r140391019
+	// Version of the manifest.
+	// ManifestVersion string `json:"manifestVersion,omitempty" yaml:"manifestVersion,omitempty"`
+
+	// NamePrefix will prefix the names of all resources mentioned in the manifest
+	// including generated configmaps and secrets.
+	NamePrefix string `json:"namePrefix,omitempty" yaml:"namePrefix,omitempty"`
+
+	// Labels to add to all objects and selectors.
+	// These labels would also be used to form the selector for apply --prune
+	// Named differently than “labels” to avoid confusion with metadata for this object
+	ObjectLabels map[string]string `json:"objectLabels,omitempty" yaml:"objectLabels,omitempty"`
+
+	// Annotations to add to all objects.
+	ObjectAnnotations map[string]string `json:"objectAnnotations,omitempty" yaml:"objectAnnotations,omitempty"`
+
+	// A Base entry specifies the relative paths within the package.
+	// It could be any format that kubectl -f allows, i.e. files, directories, URLs and globs.
+	Bases []string `json:"bases,omitempty" yaml:"bases,omitempty"`
+
+	// An Overlay entry is very similar to an Base entry.
+	// It specifies the relative paths within the package, and could be any format that kubectl -f allows.
+	// It should be able to be merged by Strategic Merge Patch on top of its corresponding Base if any.
+	Overlays []string `json:"overlays,omitempty" yaml:"overlays,omitempty"`
+
+	// List of configmaps to generate from configuration sources.
+	// Base/overlay concept doesn't apply to this field.
+	// If a configmap want to have a base and an overlay, it should go to Bases and Overlays fields.
+	Configmaps []ConfigMap `json:"configmaps,omitempty" yaml:"configmaps,omitempty"`
+
+	// List of secrets to generate from secret sources.
+	// Base/overlay concept doesn't apply to this field.
+	// If a secret want to have a base and an overlay, it should go to Bases and Overlays fields.
+	Secrets []Secret `json:"secrets,omitempty" yaml:"secrets,omitempty"`
+
+	// Whether PersistentVolumeClaims should be deleted with the other resources.
+	OwnPersistentVolumeClaims bool `json:"ownPersistentVolumeClaims,omitempty" yaml:"ownPersistentVolumeClaims,omitempty"`
+
+	// Whether recursive look for Kube-manifest.yaml, similar to `kubectl --recursive` behavior.
+	Recursive bool `json:"recursive,omitempty" yaml:"recursive,omitempty"`
+
+	// Whether prune resources not defined in Kube-manifest.yaml, similar to `kubectl apply --prune` behavior.
+	Prune bool `json:"prune,omitempty" yaml:"prune,omitempty"`
+}
```

Comment:
how manifest and package resources linked together ?

Reply:
They are not explicitly linked at present. They have different consumers. The Package information would be consumed by registry tools. The Manifest information would be consumed by kubectl and/or other configuration tools. I'd expect both to be present in the same directory tree. However, in the presence of overlays, nested packages, and other scenarios, there would be multiple Manifest instances in the tree.

Reply:
Also note that they don't have to be instantiated as resources. But are structured such that they could be.

Reply:
I started several comments then deleted them all
as i realized they were trying to paper over
trouble with `Package` and `Manifest`
per my own experience with other manifest schemes.

Some examples of wildey used manifests: [android](https://developer.android.com/guide/topics/manifest/manifest-intro.html), [chrome](https://developer.chrome.com/extensions/manifest), [microsoft](https://msdn.microsoft.com/en-us/library/windows/desktop/aa375635(v=vs.85).aspx), and apple [plist file](https://developer.xamarin.com/guides/ios/deployment,_testing,_and_metrics/app_distribution/itunesmetadata/
) (generally hidden by Xcode). Each contain

* app-descriptor fields - icon, prose description, version, etc.
* app-operation fields - activities, services, etc.

Observations:

1. If app-operation fields are deleted, app-descriptor fields serve no purpose.
   Both are components of the manifest, not independent things, even though they
   are obviously consumed differently.  Likewise, an app-descriptor is never
   attached to multiple sets of app-operation fields.  DAM is no different.
   Separating app-descriptor fields into something with a distinct lifecycle
   only creates a disassociation problem.

2. Obviously these manifests describe an app "instance".  Snapchat's manifest is
   not intended as a template from which to create new chat apps.  The DAM
   manifest, OTOH wants to deliver information allowing developers to maintain N
   different instances.  Nevertheless, each instance is rooted in a cloned base
   that is periodically rebased.  One manifest still makes sense.

Pondering the manifests already out there, i suggest these changes:

1. Follow convention - make `Manifest` the top level thing.

1. Description fields should stay in the `Manifest`.

   It's fine to gather them in a struct for convenience, but call that
   struct something that doesn't cognitively collide with Go and Java's
   `package`.

1. `Manifest` and `Overlay` are different things.

   K8s Config is easy to explain in linear algebra terms if we present a
   manifest as a state vector, and an overlay as a matrix to apply to that
   vector.  The meaning of an overlay, and the possiblity of chaining them
   becomes obvious.

   Brian's initial proposal to have two different types is less confusing and
   easier to validate.

1. Simplify file layout.

   Reasons discussed below in terms of development lifecycle.

1. Drop the `recursive` field.

   The notion of a file tree with Manifest and Overlay files sprinkled around in
   it needs clear explantion else it should be treated as bad input.  For
   example, in what order does one apply disovered overlays?  How does one avoid
   surprises (e.g. like the surprise --prune already provides).


Structs:
```
  // Manifest contains everything needed to create a k8s cluster.
  type Manifest struct {
     <standard k8s resource fields>

     // Metadata about the app.   <-- name "metadata" taken by k8s.  Eschew
     descriptor Descriptor        <-- "Package" because GoLang cognitive overload.

     // Paths (relative to manifest location) to k8s resource YAML files
     // that constitute the initial cluster state.  The files are present
     // in the manifest distribution, and can be recovered if misplaced by,
     // say, git checkout.  Other files found in the same tree yet not
     // specified here may belong to overlays, or may simply be added to
     // the cluster.
     Resources []string

     // Generators of ConfigMap and Secret instances from env files, etc.
     ConfigMapGenerators []ConfigMapGenerator
     SecretGenerators []SecretGenerator
  }

  // Descriptor is a convenience struct gathering metadata to support
  // Manifest search, browse and updates.
  type Descriptor struct {
     <standard k8s resource fields>
     Name string
     // Change version when a resource added, but not, say when a container
     // image changes. This field maintained by the manifest author.
     Version string
     Description string
     Icon string
     Keywords []string
     Homepage string
     ...
  }

  // Overlay describes changes to make to a Manifest or the
  // result of an Overlay applied to a Manifest.
  type Overlay struct {
     <standard k8s resource fields>

     // Evolves separately from Manifest version.
     // This field maintained by the owner of the overlay.
     Version string

     // Path to either another Overlay or a Manifest file.
     // Clients like kubectl will follow these links.
     // If unspecified, defaults to './Kube-manifest.yaml'
     Operand string

     NamePrefix string
     ObjectLabels map...
     ObjectAnnotations map...

     // Resources to add.  Delete a resource by just deleting it from the tree.
     Resources []string

     // Files consumable as a strategic-merge-patch.
     Smps []string

     OwnPvc bool
     Prune bool  // needs clear explanation
     // Recursive bool -- Dropped until we can agree what it means.
  }
```

Then this simple arrangement in the Manifest tarball/githubrepo:
```
  Kube-manifest.yaml   // Instance of Manifest
  LICENSE
  README.md
  identity-overlay.yaml // Provided as a convenience - a NO-OP overlay.
  resources/
    deployment.yaml
    ingress.yaml
```

Finally (no different from Brian's plan), the end user
1. Clones the upstream repo after finding it in the store.
1. Edits `Kube-manifest.yaml` to suit their purposes.
1. Optionally renames `identity-overlay.yaml` to
   `my-overlay.yaml` and adds any resources the overlay
   needs to the `resources/` directory.
1. Checks everything in to their own repo for
   periodic rebasing against upstream.
1. Runs `kubectl apply -f my-overlay.yaml`

Reply:
> identity-overlay.yaml // Provided as a convenience - a NO-OP overlay.

I think this tree structure may encourage the users to put a bunch of overlay k8s API resources in one single file, which may not be a good practice to manage different kinds of resources.

Reply:
@monopole 

Thanks for the manifest examples.

I like "Descriptor" for the package metadata. I wouldn't want an application registry to look at the other manifest fields, though.

Embedded structures shouldn't generally contain K8s API resource fields. Do you envision generating CRDs from those embedded resources?

One usage model that I'd like to support is the following:
* Fork a package that contains base resources
* Create multiple variants, implemented using overlays, additional resources, and variant-specific name prefixes and labels

Another is:
* Fork a package that contains base resources
* Customize it by mutating the base resources, adding additional resources (e.g., secrets), and specifying a name prefix and labels
* It should be possible to easily rebase to upgrade the bases

Another is:
* I just want to create declarative instructions for kubectl so that I can update my app with one command instead of a shell script, and make it impossible to forget to specify the appropriate label selector with apply --prune

It is common in Borg to put directories for variants and the directory for the base templates at the same level in the directory hierarchy. That could work. But I think it get confusing for files corresponding to multiple variants to go into the same subdirectory. Either way, though we should recommend a convention, I think we should support syntax that leaves the structure up to the user.

I'm fine with omitting the recursive option for now.

Reply:
In your proposal, an overlay is always required, correct?

Also, the configmap and secret generators would need to be added to Overlay.

Reply:
> Do you envision generating CRDs from those resources?

Yes,
`PackageMetadata` could be used by other workflow managers. It doesn't have to refer exclusively to this  new `Manifest` type. I would keep app-operation and app-descriptor separated.

Can we reference `PackageMetadata` resource from the Manifest?
```
     descriptor: String       <-- " PackageResource name”
```

> Manifest and Overlay are different things.

How are they different and what features would it add to separate them ? 

IMO, in your proposal `Overlays` could be seen as a `Manifest` instance.
In that case, having an `Overlay` (even empty) is a requirement and it became the main resource in the cluster.
```
kubectl apply -f overlay.yaml
kubectl get overlays
kubectl delete overlay foo #  cascade delete
```
I’d prefer to keep them optional and only a way to declare small variations when templating / transformations aren’t enough.

> the possiblity of chaining them becomes obvious.

Do we need to support chaining overlays, it would increase complexity for the user and become more fragile.  
From Brian’s [declarative app](https://docs.google.com/document/d/1cLPGweVEYrVqQvBLJg6sxV-TrE5Rm2MNOBA_cxZP2WU/edit#heading=h.dr88tktf0e99): 
> I do not think we should support arbitrary numbers of stacked overlays. [...] changing a base layer invalidates layers stacked on top of it. Beyond a couple layers it also becomes hard to reason about the implications of each overlay.

Reply:
@ant31 There are detailed examples of why Manifest and Overlay are different.  The Manifest is a state you download and edit as desired, the Overlay is an operator - it drives the cluster represented by the manifest to a new state _via mechanisms that cannot be achieved by editing the manifest directly_.  It's also the way you get to N variations on the base state.  One editted manifest is just one state (N==1).  A manifest plus one overlay is still just N==1.  Also, the Manifest is registered in a store, the Overlay is not.  One *could* register the overlay in a store, but it wouldn't be an independent thing - it points to a manifest.

@bgrant0607 and @ant31 - Yes, this supports those three goals.

> Fork a package that contains base resources. Create multiple variants..
Fork the repo for the manifest, create overlay1, overlay2, etc. Run 
* `kubectl apply -f overlay1.yaml`  // internal to this file is a pointer to Kube-manifest.xml
* `kubectl apply -f overlay2.yaml`  // ditto
* ...

> Fork a package, customize it by mutating the base resources....

Fork the repo for the manifest, edit as desired, add new resources, don't create overlays. Run 
* `kubectl apply -f Kube-Manifest.yaml`

> I just want to create declarative instructions for kubectl 

Create a manifest and apply it.

@bgrant0607 
> In your proposal, an overlay is always required, correct?

To get the name-prefix, alas, yes.  We could embed the "first" overlay directly in the manifest, and i initially wrote it up that way, and then found that the whole thing was much harder to explain.  You have to say stuff like:  _The overlay embedded in the manifest is applied before any overlays from the outside are applied._  weird.

Brian's right to presume one wouldn't chain overlays in practice.  However, we either create a means to do it sensibly, or we should prohibit it entirely.  Clearly the manifest is a state, clearly the overlay modifies that state.  You almost have to go out of your way to prohibit the application of another overlay.

Reply:
OK, Brian and I have had another chat.  I think my objections have been addressed, and i can get behind conflating what i was calling Manifest and Overlay, and that @ant31 is right in claiming they are the same(-ish).  `Descriptor` the name seems to have survived the discussion as a replacement for `Package`, and we can add a field to Manifest to allow one to find its Descriptor.  We'll also add more description to pin down how one, when presented with a directory tree containing manifest files, arrives at  a single unambiguous order of expansion.